### ISSUE TITLE
Java VerifyError with specific Boolean Let variable

### ISSUE THREAD

**Issue by rdwebster** [levans/Open-Quark#4]
The following code produces a Java VerifyError when executed:

```
xyz :: Boolean -> Boolean -> String;
xyz !flag1 !flag2 = 
    let
        condition :: Boolean;
        condition = flag1 || Prelude.not flag2;
    in
        if condition then
            "abc"
        else
            "def";
```

The Java error:
    java.lang.VerifyError: (class: org/openquark/cal_Com_Indicee_Imp_Model_ExpensesDataModel/Xyz, method: condition$3$def_Lazy signature: (ZZLorg/openquark/cal/internal/runtime/lecc/RTExecutionContext;)Lorg/openquark/cal/internal/runtime/lecc/RTValue;) Expecting to find object/array on stack,  Detail: (class: org/openquark/cal_Com_Indicee_Imp_Model_ExpensesDataModel/Xyz, method: condition$3$def_Lazy signature: (ZZLorg/openquark/cal/internal/runtime/lecc/RTExecutionContext;)Lorg/openquark/cal/internal/runtime/lecc/RTValue;) Expecting to find object/array on stack

With LECC source enabled, the Java code generated does not compile:
    Error: Failed to finalize Java code for: Com.Indicee.Imp.Model.ExpensesDataModel Detail: Error compiling generated source for module Com.Indicee.Imp.Model.ExpensesDataModel.
    /Users/rdwebster/Dev/Indicee/Helper_Quark_Binaries/bin/lecc_runtime/org/openquark/cal_Com_Indicee_Imp_Model_ExpensesDataModel/Xyz.java:58: cannot find symbol
    symbol  : constructor RTAppS(org.openquark.cal_Cal_Core_Prelude.Or,boolean,boolean)
    location: class org.openquark.cal_Cal_Core_Prelude.Or.RTAppS
            return new Or.RTAppS(Or.$instance, flag1, !flag2);

Removing the pling on the 2nd arg or in-lining the condition fixes the problem.

**Comment by indicee-michald** [comment]
Could not reproduce (prior to any of my other bug fixes).

**Comment by rdwebster** [comment]
I searched my email and found out why this isn't reproducible anymore -- it has been fixed!
Ray provided a fix for this in Aug 2010, but the bug wasn't marked as closed.

Here are some notes he provided, along with info for a possible optimization...

Basically what was happening was we were applying an optimization of Prelude.Not when the argument was known to be evaluated in a situation where the result needs to be boxed but we weren't doing the boxing.  It was definitely a bug.  But it does bring up the fact that we're missing an optimization.  With the bug fixed we're generating:
    private static final RTValue condition$3$def_Lazy(boolean flag1, boolean flag2, RTExecutionContext $ec) throws CALExecutorException {
        return 
            new Or.RTAppS(Or.$instance, flag1, RTData.CAL_Boolean.make(!flag2));
    }
While this is correct we could be optimizing to: 
    private static final RTValue condition$3$def_Lazy(boolean flag1, boolean flag2, RTExecutionContext $ec) throws CALExecutorException {
        return 
            new RTData.CAL_Boolean.make(flag1 || (!flag2));
    }
If I get some spare time (ha, ha, ha, sometimes I crack myself up) I'll look into why we're not optimizing in this case.

Anyway the change is on line 6578 in the buildStrictApplicationNode() method.
